prov/verbs/src/verbs_mr.c

```diff
@@ -88,79 +164,138 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
 	*mr = &md->mr_fid;
-	if (md->domain->eq_flags & FI_REG_MR) {
-		struct fi_eq_entry entry = {
-			.fid = &md->mr_fid.fid,
-			.context = context
-		};
-		if (md->domain->eq)
-			fi_ibv_eq_write_event(md->domain->eq, FI_MR_COMPLETE,
-				 	      &entry, sizeof(entry));
-		else if (md->domain->util_domain.eq)
-			 /* This branch is taken for the verbs/DGRAM */
-			fi_eq_write(&md->domain->util_domain.eq->eq_fid,
-				    FI_MR_COMPLETE, &entry, sizeof(entry), 0);
-	}
-	return 0;
-
+	return FI_SUCCESS;
 err:
 	free(md);
-	return -errno;
+	return -ret;
+}
+
+FI_IBV_DEFINE_MR_REG_OPS(_)
+
+static int fi_ibv_mr_mgr_close(fid_t fid)
+{
+	struct fi_ibv_mem_desc *md =
+		container_of(fid, struct fi_ibv_mem_desc, mr_fid.fid);
+	ofi_mr_mgr_remove(&md->domain->mgr,
+			  md->entry);
+
+	return FI_SUCCESS;
 }
 
-static int fi_ibv_mr_regv(struct fid *fid, const struct iovec * iov,
-		size_t count, uint64_t access, uint64_t offset, uint64_t requested_key,
-		uint64_t flags, struct fid_mr **mr, void *context)
+static struct fi_ops fi_ibv_mr_mgr_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = fi_ibv_mr_mgr_close,
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+int fi_ibv_monitor_subscribe(struct ofi_mem_monitor *notifier, void *addr,
+			     size_t len, struct ofi_subscription *subscription)
 {
-	if (count > VERBS_MR_IOV_LIMIT) {
+	return FI_SUCCESS;
+}
+
+void fi_ibv_monitor_unsubscribe(struct ofi_mem_monitor *notifier, void *addr,
+				size_t len, struct ofi_subscription *subscription)
+{
+}
+
+struct ofi_subscription *fi_ibv_monitor_get_event(struct ofi_mem_monitor *notifier)
+{
+	return NULL;
+}
+
+int fi_ibv_mr_cache_entry_reg(struct ofi_mr_cache *cache,
+			      struct ofi_mr_cache_entry *entry,
+			      void *address, size_t length)
+{
+	int fi_ibv_access = IBV_ACCESS_LOCAL_WRITE |
+			    IBV_ACCESS_REMOTE_WRITE |
+			    IBV_ACCESS_REMOTE_ATOMIC |
+			    IBV_ACCESS_REMOTE_READ;
+	struct fi_ibv_mem_desc *md = (struct fi_ibv_mem_desc *)entry->data;
+	md->domain = container_of(cache->domain, struct fi_ibv_domain, util_domain);
+	md->mr_fid.fid.ops = &fi_ibv_mr_mgr_ops;
+	return fi_ibv_mr_reg_common(md, fi_ibv_access, address, length,
+				    entry->reg_attr.context);
+}
+
+int fi_ibv_mr_cache_entry_dereg(struct ofi_mr_cache *cache,
+				struct ofi_mr_cache_entry *entry)
+{
+	struct fi_ibv_mem_desc *md = (struct fi_ibv_mem_desc *)entry->data;
+
+	return -ibv_dereg_mr(md->mr);
+}
+
+static int fi_ibv_mr_mgr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
+				 uint64_t flags, struct fid_mr **mr)
+{
+	struct fi_ibv_domain *domain;
+	struct ofi_mr_mgr_entry *entry;
+	struct fi_ibv_mem_desc *md;
+
+	if (OFI_UNLIKELY(flags))
+		return -FI_EBADFLAGS;
+	if (OFI_UNLIKELY(fid->fclass != FI_CLASS_DOMAIN))
+		return -FI_EINVAL;
+	if (OFI_UNLIKELY(attr->iov_count > VERBS_MR_IOV_LIMIT)) {
 		VERBS_WARN(FI_LOG_FABRIC,
 			   "iov count > %d not supported\n",
 			   VERBS_MR_IOV_LIMIT);
 		return -FI_EINVAL;
 	}
```

Comment:
The checks above seem to be common for both this function and fi_ibv_mr_regattr. They can be moved into a check function to be called from both.

Reply:
Done